private/bufpkg/bufmodule/bufmodulecache/cas_module_cacher.go

```diff
@@ -0,0 +1,272 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bufmodulecache
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule"
+	"github.com/bufbuild/buf/private/bufpkg/bufmodule/bufmoduleref"
+	"github.com/bufbuild/buf/private/pkg/manifest"
+	"github.com/bufbuild/buf/private/pkg/normalpath"
+	"github.com/bufbuild/buf/private/pkg/storage"
+	"go.uber.org/multierr"
+	"go.uber.org/zap"
+)
+
+// subdirectories under ~/.cache/buf/v2/{remote}/{owner}/{repo}
+const (
+	blobsDir   = "blobs"
+	commitsDir = "commits"
+)
+
+type casModuleCacher struct {
+	logger *zap.Logger
+	bucket storage.ReadWriteBucket
+}
+
+var _ moduleCache = (*casModuleCacher)(nil)
+
+func (c *casModuleCacher) GetModule(
+	ctx context.Context,
+	modulePin bufmoduleref.ModulePin,
+) (_ bufmodule.Module, retErr error) {
+	moduleBasedir := normalpath.Join(modulePin.Remote(), modulePin.Owner(), modulePin.Repository())
+	manifestDigestStr := modulePin.Digest()
+	if manifestDigestStr == "" {
+		// Attempt to look up manifest digest from commit
+		commitPath := normalpath.Join(moduleBasedir, commitsDir, modulePin.Commit())
+		manifestDigestBytes, err := c.loadPath(ctx, commitPath)
+		if err != nil {
+			return nil, err
+		}
+		manifestDigestStr = string(manifestDigestBytes)
+	}
+	manifestDigest, err := manifest.NewDigestFromString(manifestDigestStr)
+	if err != nil {
+		return nil, err
+	}
+	manifestFromCache, err := c.readManifest(ctx, moduleBasedir, manifestDigest)
+	if err != nil {
+		return nil, err
+	}
+	var blobs []manifest.Blob
+	blobDigests := make(map[string]struct{})
+	for _, path := range manifestFromCache.Paths() {
+		digest, found := manifestFromCache.DigestFor(path)
+		if !found {
+			return nil, fmt.Errorf("digest not found for path: %s", path)
+		}
+		if _, ok := blobDigests[digest.String()]; ok {
+			// We've already loaded this blob
+			continue
+		}
+		blob, err := c.readBlob(ctx, moduleBasedir, digest)
+		if err != nil {
+			return nil, err
+		}
+		blobs = append(blobs, blob)
+	}
+	blobSet, err := manifest.NewBlobSet(ctx, blobs)
+	if err != nil {
+		return nil, err
+	}
+	return bufmodule.NewModuleForManifestAndBlobSet(ctx, manifestFromCache, blobSet)
+}
+
+func (c *casModuleCacher) PutModule(
+	ctx context.Context,
+	modulePin bufmoduleref.ModulePin,
+	module bufmodule.Module,
+) (retErr error) {
+	moduleManifest := module.Manifest()
+	manifestBlob, err := moduleManifest.Blob()
+	if err != nil {
+		return err
+	}
```

Comment:
Nit: You documented that `.Manifest` could be nil, you could check that here before asking its `.Blob` to avoid nil panic?

Reply:
We check it a layer higher in the cas_module_reader, but I added an extra check here.